subprojects/version-control/src/testFixtures/groovy/org/gradle/vcs/fixtures/TemporaryGitRepository.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.vcs.fixtures;
+
+import org.eclipse.jgit.dircache.DirCacheEntry;
+import org.eclipse.jgit.internal.storage.file.FileRepository;
+import org.eclipse.jgit.junit.RepositoryTestCase;
+import org.eclipse.jgit.lib.FileMode;
+import org.eclipse.jgit.lib.ObjectId;
+import org.eclipse.jgit.lib.Repository;
+import org.eclipse.jgit.revwalk.RevCommit;
+import org.eclipse.jgit.treewalk.FileTreeIterator;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+
+public class TemporaryGitRepository extends RepositoryTestCase implements TestRule {
```

Comment:
I think we might be able to get away with less here by abstracting away things a bit.  I don't think we need the same level of control/detail that the JGit unit tests need.  It seems like some of the functionality provided here already exists in a more familiar form in our other tests (e.g., `TestFile`).

I think for our use, we need a test fixture that just does something like...create a temporary directory that's relative to the current test that we can use as a source repository and then initialize the repository (this can happen in the Rule).

From the fixture, we then need to be able to get a URL to the repo, commit, create a branch, create a tag.  These methods can have assumptions about how they behave.  e.g., `createBranch(branchName)` can automatically check you out to that branch. Commit could just be `commit(message, files)`.

The way I would expect to see this used in tests would be something like:
```
@Rule TemporaryGitRepository git
def "some test"() {
   def sourceFile = file("source")
   sourceFile.text = "asdf"
   git.commit("initial", sourceFile)
   def gitVcs = new GitVersionControlSystem()
   // pseudo:
   def spec = create spec with git.url
   when:
   gitVcs.populate(workingDir, spec)
   then:
   workingDir.file("source").text == "asdf"
}
```

This is close to what you have in your test, but I think we should try to use our own test fixture vs JGit directly since we can simplify/give better names.  And if we absolutely must, provide an escape hatch to get to a `Git` in the fixture.

Reply:
I've given this a stab. But, I think I may still not have made it as simple as you are suggesting. But, I think it makes sense that we have to know the working tree of the temporary git repository being managed by the `TemporaryGitRepository` to know where to add the files we want to commit to that repository.